lxd/device/gpu_physical.go

```diff
@@ -98,10 +101,202 @@ func (d *gpuPhysical) Start() (*deviceConfig.RunConfig, error) {
 	return d.startContainer()
 }
 
+// startCDIDevices starts all the devices given in a CDI specification:
+// * `unix-char` (representing the card and non-card devices)
+// * `disk` (representing the mounts)).
+func (d *gpuPhysical) startCDIDevices(cdiID cdi.ID, configDevices cdi.ConfigDevices, runConf *deviceConfig.RunConfig) error {
+	srcFDHandlers := make([]*os.File, 0)
+	defer func() {
+		for _, f := range srcFDHandlers {
+			_ = f.Close()
+		}
+	}()
+
+	for _, conf := range configDevices.UnixCharDevs {
+		if conf["source"] == "" {
+			return fmt.Errorf("The source of the unix-char device %q used for CDI is empty", conf)
+		}
+
+		_, major, minor, err := unixDeviceAttributes(conf["source"])
+		if err != nil {
+			return err
+		}
+
+		// Here putting a 'cdi.unix' prefix with 'd.name' as a device name with create an directory entry like:
+		// <lxd_var_path>/devices/<instance_name>/cdi.unix.<gpu_device_name>.<path_encoded_relative_dest_path>
+		// 'unixDeviceSetupCharNum' is already checking for dupe entries so we have no validation to do here.
+		err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "cdi.unix", d.name, conf, major, minor, conf["path"], false, runConf)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, conf := range configDevices.BindMounts {
+		if conf["source"] == "" {
+			return fmt.Errorf("The source of the disk device %q used for CDI is empty", conf)
+		}
+
+		srcPath := shared.HostPath(conf["source"])
+		destPath := conf["path"]
+		relativeDestPath := strings.TrimPrefix(destPath, "/")
+
+		// This time, the created path will be like:
+		// <lxd_var_path>/devices/<instance_name>/cdi.disk.<gpu_device_name>.<path_encoded_relative_dest_path>
+		deviceName := filesystem.PathNameEncode(deviceJoinPath("cdi.disk", d.name, relativeDestPath))
+		devPath := filepath.Join(d.inst.DevicesPath(), deviceName)
+
+		ownerShift := deviceConfig.MountOwnerShiftNone
+		if shared.IsTrue(conf["shift"]) {
+			ownerShift = deviceConfig.MountOwnerShiftDynamic
+		}
+
+		options := []string{"bind"}
+		mntOptions := shared.SplitNTrimSpace(conf["raw.mount.options"], ",", -1, true)
+		fsName := "none"
+
+		fileInfo, err := os.Stat(srcPath)
+		if err != nil {
+			return fmt.Errorf("Failed accessing source path %q: %w", srcPath, err)
+		}
+
+		fileMode := fileInfo.Mode()
+		isFile := false
+		if !fileMode.IsDir() {
+			isFile = true
+		}
+
+		f, err := os.OpenFile(srcPath, unix.O_PATH|unix.O_CLOEXEC, 0)
+		if err != nil {
+			return fmt.Errorf("Failed opening source path %q: %w", srcPath, err)
+		}
+
+		srcPath = fmt.Sprintf("/proc/self/fd/%d", f.Fd())
+		srcFDHandlers = append(srcFDHandlers, f)
+
+		// Create the devices directory if missing.
+		if !shared.PathExists(d.inst.DevicesPath()) {
```

Comment:
do we need this on every iteration or can it be done once at the top if devices > 0?

Reply:
Yes we can move this on top actually! Thanks

Reply:
fixed